File: cli/internal/connectapi/admin_event_log.go
package connectapi

import (
	"context"
	"time"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/types/known/timestamppb"
	"hmans.de/chatto/internal/core"
	adminv1 "hmans.de/chatto/internal/pb/chatto/admin/v1"
)

type adminEventLogService struct {
	api *API
}

func (s *adminEventLogService) ListEvents(ctx context.Context, req *connect.Request[adminv1.ListEventsRequest]) (*connect.Response[adminv1.ListEventsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	filter, err := adminEventLogFilterFromAPI(req.Msg.GetFilter())
	if err != nil {
		return nil, err
	}
	conn, err := s.api.core.ListEventLog(ctx, caller.UserID, core.EventLogQuery{
		Limit:  int(req.Msg.GetLimit()),
		Before: req.Msg.GetBefore(),
		Filter: filter,
	})
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(adminEventLogConnectionToAPI(conn)), nil
}

func (s *adminEventLogService) ListEventTypes(ctx context.Context, _ *connect.Request[adminv1.ListEventTypesRequest]) (*connect.Response[adminv1.ListEventTypesResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	eventTypes, err := s.api.core.EventLogEventTypes(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&adminv1.ListEventTypesResponse{EventTypes: eventTypes}), nil
}

func (s *adminEventLogService) GetEvent(ctx context.Context, req *connect.Request[adminv1.GetEventRequest]) (*connect.Response[adminv1.GetEventResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	entry, err := s.api.core.GetEventLogEntry(ctx, caller.UserID, req.Msg.GetSequence())
	if err != nil {
		return nil, connectError(err)
	}
	if entry == nil {
		return nil, connectError(core.ErrNotFound)
	}
	return connect.NewResponse(&adminv1.GetEventResponse{
		Entry: adminEventLogEntryToAPI(entry),
	}), nil
}

func adminEventLogFilterFromAPI(filter *adminv1.AdminEventLogFilter) (core.EventLogFilter, error) {
	if filter == nil {
		return core.EventLogFilter{}, nil
	}
	from, err := checkedEventLogTimestamp(filter.GetCreatedAtFrom(), "created_at_from")
	if err != nil {
		return core.EventLogFilter{}, err
	}
	to, err := checkedEventLogTimestamp(filter.GetCreatedAtTo(), "created_at_to")
	if err != nil {
		return core.EventLogFilter{}, err
	}
	return core.EventLogFilter{
		EventType:     filter.GetEventType(),
		ActorID:       filter.GetActorId(),
		CreatedAtFrom: from,
		CreatedAtTo:   to,
	}, nil
}

func checkedEventLogTimestamp(ts *timestamppb.Timestamp, field string) (*time.Time, error) {
	if ts == nil {
		return nil, nil
	}
	if err := ts.CheckValid(); err != nil {
		return nil, invalidArgument(field + " is invalid")
	}
	value := ts.AsTime()
	return &value, nil
}

func adminEventLogConnectionToAPI(conn *core.EventLogConnection) *adminv1.ListEventsResponse {
	if conn == nil {
		return &adminv1.ListEventsResponse{}
	}
	entries := make([]*adminv1.AdminEventLogEntry, 0, len(conn.Entries))
	for _, entry := range conn.Entries {
		entries = append(entries, adminEventLogEntryToAPI(entry))
	}
	return &adminv1.ListEventsResponse{
		Entries:      entries,
		HasOlder:     conn.HasOlder,
		EndCursor:    conn.EndCursor,
		TotalCount:   conn.TotalCount,
		ScannedCount: conn.ScannedCount,
		ScanLimit:    conn.ScanLimit,
		ScanLimited:  conn.ScanLimited,
	}
}

func adminEventLogEntryToAPI(entry *core.EventLogEntry) *adminv1.AdminEventLogEntry {
	if entry == nil {
		return nil
	}
	return &adminv1.AdminEventLogEntry{
		Sequence:      entry.Sequence,
		Subject:       entry.Subject,
		AggregateType: entry.AggregateType,
		AggregateId:   entry.AggregateID,
		EventType:     entry.EventType,
		EventId:       entry.EventID,
		ActorId:       entry.ActorID,
		CreatedAt:     entry.CreatedAt,
		PayloadJson:   entry.PayloadJSON,
	}
}

File: apps/frontend/src/lib/api/adminEventLog.ts
import { Timestamp } from '@bufbuild/protobuf';
import { Code, ConnectError, createClient } from '@connectrpc/connect';
import { createConnectTransport } from '@connectrpc/connect-web';
import { AdminEventLogService } from '$lib/pb/chatto/admin/v1/event_log_connect';
import type { AdminEventLogEntry as APIAdminEventLogEntry } from '$lib/pb/chatto/admin/v1/event_log_pb';

export type AdminEventLogAPIConfig = {
  baseUrl: string;
  bearerToken: string | null;
};

export type AdminEventLogFilter = {
  eventType: string;
  actorId: string;
  createdAtFrom: string;
  createdAtTo: string;
};

export type AdminEventLogEntry = {
  sequence: string;
  subject: string;
  aggregateType: string;
  aggregateId: string;
  eventType: string;
  eventId: string;
  actorId: string;
  createdAt: string;
  payloadJson: string;
};

export type AdminEventLogPage = {
  entries: AdminEventLogEntry[];
  hasOlder: boolean;
  endCursor: string | null;
  totalCount: string;
  scannedCount: number;
  scanLimit: number;
  scanLimited: boolean;
};

export type AdminEventLogAPI = ReturnType<typeof createAdminEventLogAPI>;

export const EMPTY_ADMIN_EVENT_LOG_FILTER: AdminEventLogFilter = {
  eventType: '',
  actorId: '',
  createdAtFrom: '',
  createdAtTo: ''
};

export function createAdminEventLogAPI(config: AdminEventLogAPIConfig) {
  const transport = createConnectTransport({
    baseUrl: config.baseUrl,
    useBinaryFormat: true
  });
  const client = createClient(AdminEventLogService, transport);
  const headers = () =>
    config.bearerToken ? { Authorization: `Bearer ${config.bearerToken}` } : undefined;

  return {
    async listEvents(input: {
      limit: number;
      before?: string | null;
      filter?: AdminEventLogFilter;
    }): Promise<AdminEventLogPage> {
      const response = await client.listEvents(
        {
          limit: input.limit,
          before: input.before ?? undefined,
          filter: eventLogFilterInput(input.filter ?? EMPTY_ADMIN_EVENT_LOG_FILTER)
        },
        { headers: headers() }
      );
      return {
        entries: response.entries.map(adminEventLogEntry),
        hasOlder: response.hasOlder,
        endCursor: response.endCursor ?? null,
        totalCount: String(response.totalCount),
        scannedCount: response.scannedCount,
        scanLimit: response.scanLimit,
        scanLimited: response.scanLimited
      };
    },

    async listEventTypes(): Promise<string[]> {
      const response = await client.listEventTypes({}, { headers: headers() });
      return [...response.eventTypes];
    },

    async getEvent(sequence: string): Promise<AdminEventLogEntry | null> {
      try {
        const response = await client.getEvent({ sequence }, { headers: headers() });
        return response.entry ? adminEventLogEntry(response.entry) : null;
      } catch (error) {
        if (error instanceof ConnectError && error.code === Code.NotFound) return null;
        throw error;
      }
    }
  };
}

function eventLogFilterInput(filter: AdminEventLogFilter) {
  if (!hasActiveEventLogFilter(filter)) return undefined;
  return {
    eventType: filter.eventType || undefined,
    actorId: filter.actorId || undefined,
    createdAtFrom: timestampFromISO(filter.createdAtFrom),
    createdAtTo: timestampFromISO(filter.createdAtTo)
  };
}

function hasActiveEventLogFilter(filter: AdminEventLogFilter): boolean {
  return Boolean(filter.eventType || filter.actorId || filter.createdAtFrom || filter.createdAtTo);
}

function timestampFromISO(value: string): Timestamp | undefined {
  if (!value) return undefined;
  return Timestamp.fromDate(new Date(value));
}

function adminEventLogEntry(entry: APIAdminEventLogEntry): AdminEventLogEntry {
  return {
    sequence: entry.sequence,
    subject: entry.subject,
    aggregateType: entry.aggregateType,
    aggregateId: entry.aggregateId,
    eventType: entry.eventType,
    eventId: entry.eventId,
    actorId: entry.actorId,
    createdAt: entry.createdAt?.toDate().toISOString() ?? '',
    payloadJson: entry.payloadJson
  };
}
